File: app/src/main/java/com/androidevlinux/percy/UTXO/adapters/NewsAdapter.java
package com.androidevlinux.percy.UTXO.adapters;

import android.content.Context;
import android.net.Uri;
import android.support.annotation.NonNull;
import android.support.customtabs.CustomTabsIntent;
import android.support.v4.widget.SwipeRefreshLayout;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.FrameLayout;
import android.widget.TextView;

import com.androidevlinux.percy.UTXO.R;
import com.androidevlinux.percy.UTXO.data.models.newsapi.Article;
import com.makeramen.roundedimageview.RoundedImageView;
import com.squareup.picasso.Picasso;

import java.util.ArrayList;

public class NewsAdapter extends RecyclerView.Adapter<NewsAdapter.ViewHolder> {

    private ArrayList<Article> articleArrayList;
    private Context context;
    private SwipeRefreshLayout mSwipeRefreshLayout;

    public NewsAdapter(ArrayList<Article> articleArrayList, Context context, SwipeRefreshLayout mSwipeRefreshLayout) {
        this.context = context;
        this.mSwipeRefreshLayout = mSwipeRefreshLayout;
        this.articleArrayList = articleArrayList;
    }

    @NonNull
    @Override
    public NewsAdapter.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View v = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.news_fragment_adapter_row, parent, false);
        return new NewsAdapter.ViewHolder(v);
    }

    @Override
    public void onBindViewHolder(@NonNull final NewsAdapter.ViewHolder holder, int position) {
        Article article = articleArrayList.get(position);
        holder.title.setText(article.getTitle());
        holder.txt_description.setText(article.getDescription());
        holder.inner_frame.setOnClickListener(view -> {
            CustomTabsIntent.Builder builder = new CustomTabsIntent.Builder();
            CustomTabsIntent customTabsIntent = builder.build();
            customTabsIntent.launchUrl(context, Uri.parse(article.getUrl()));
        });
        if (null != article.getUrlToImage() && !article.getUrlToImage().isEmpty()) {
            Picasso.get().load(article.getUrlToImage()).into(holder.articleImage);
        }
        holder.txt_author.setText(article.getAuthor());
        holder.txt_date.setText(article.getPublishedAt());

        if (articleArrayList.size() == 20) {
            if (mSwipeRefreshLayout.isRefreshing()) {
                mSwipeRefreshLayout.setRefreshing(false);
            }
        }
    }

    @Override
    public int getItemCount() {
        return articleArrayList.size();
    }

    /**
     * View holder class
     */
    class ViewHolder extends RecyclerView.ViewHolder {
        private FrameLayout inner_frame;
        private RoundedImageView articleImage;
        private TextView title;
        private TextView txt_description;
        private TextView txt_author;
        private TextView txt_date;

        private ViewHolder(View view) {
            super(view);
            inner_frame = view.findViewById(R.id.inner_frame);
            articleImage = view.findViewById(R.id.discover_icon);
            title = view.findViewById(R.id.txt_title);
            txt_description = view.findViewById(R.id.txt_description);
            txt_author = view.findViewById(R.id.txt_author);
            txt_date = view.findViewById(R.id.txt_date);
        }
    }
}

File: app/src/main/java/com/androidevlinux/percy/UTXO/adapters/NewsAdapter.kt
package com.androidevlinux.percy.UTXO.adapters

import android.content.Context
import android.net.Uri
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.animation.AnimationUtils
import android.widget.TextView
import androidx.appcompat.widget.AppCompatImageView
import androidx.browser.customtabs.CustomTabsIntent
import androidx.cardview.widget.CardView
import androidx.swiperefreshlayout.widget.CircularProgressDrawable
import com.androidevlinux.percy.UTXO.R
import com.androidevlinux.percy.UTXO.data.models.newsapi.Article
import com.squareup.picasso.Picasso
import java.text.SimpleDateFormat
import java.util.*

class NewsAdapter(private val articleArrayList: ArrayList<Article>, private val context: Context) : androidx.recyclerview.widget.RecyclerView.Adapter<NewsAdapter.ViewHolder>() {
    private var lastPosition = -1

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val v = LayoutInflater.from(parent.context)
                .inflate(R.layout.news_fragment_adapter_row, parent, false)
        return ViewHolder(v)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val article = articleArrayList[position]
        val animation = AnimationUtils.loadAnimation(
                context,
                if (position > lastPosition)
                    R.anim.up_from_bottom
                else
                    R.anim.down_from_top
        )
        holder.itemView.startAnimation(animation)
        lastPosition = holder.adapterPosition

        holder.title.text = article.title
        holder.txtDescription.text = article.description
        holder.innerFrame.setOnClickListener {
            val builder = CustomTabsIntent.Builder()
            val customTabsIntent = builder.build()
            customTabsIntent.launchUrl(context, Uri.parse(article.url))
        }
        val circularProgressDrawable = CircularProgressDrawable(context)
        circularProgressDrawable.strokeWidth = 5f
        circularProgressDrawable.centerRadius = 30f
        circularProgressDrawable.setColorSchemeColors(R.color.md_blue,
                R.color.md_pink)
        circularProgressDrawable.start()
        if (null != article.urlToImage && article.urlToImage!!.isNotEmpty()) {
            Picasso.get().load(article.urlToImage).placeholder(circularProgressDrawable).into(holder.articleImage)
        }
        val sdf = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss'Z'", Locale.getDefault())
        sdf.timeZone = TimeZone.getTimeZone("GMT")
        sdf.parse(article.publishedAt)
        val outputFormat = SimpleDateFormat("E, dd MMM yyyy HH:mm aaa", Locale.getDefault())
        val inputDateStr = article.publishedAt
        val date = sdf.parse(inputDateStr)
        val outputDateStr = outputFormat.format(date)


        holder.txtAuthor.text = article.author
        holder.txtDate.text = outputDateStr


    }

    override fun getItemCount(): Int {
        return articleArrayList.size
    }

    /**
     * View holder class
     */
    inner class ViewHolder(view: View) : androidx.recyclerview.widget.RecyclerView.ViewHolder(view) {
        val innerFrame: CardView = view.findViewById(R.id.inner_frame)
        val articleImage: AppCompatImageView = view.findViewById(R.id.news_poster)
        val title: TextView = view.findViewById(R.id.txt_title)
        val txtDescription: TextView = view.findViewById(R.id.txt_description)
        val txtAuthor: TextView = view.findViewById(R.id.txt_author)
        val txtDate: TextView = view.findViewById(R.id.txt_date)

    }
}
